pkg/core/blockchain.go

```diff
@@ -792,53 +792,86 @@ func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage)
 
 		const persistBatchSize = 200000
 		var (
-			seekErr        error
-			cnt            int
-			storageItmsCnt int
-			contractIDsCnt int
-			batchCnt       int
+			seekErr                                       uatomic.Error
+			cnt, storageItmsCnt, contractIDsCnt, batchCnt int
+			errCh                                         = make(chan error)
+			workers                                       uatomic.Int32
+			exit                                          = make(chan struct{})
 		)
-		trieStore.Seek(storage.SeekRange{Prefix: []byte{byte(oldStoragePrefix)}}, func(k, v []byte) bool {
-			if seekErr != nil {
-				return false
-			}
-			if cnt >= persistBatchSize {
-				cnt = 0
-				keys, seekErr = cache.Persist()
-				if seekErr != nil {
-					seekErr = fmt.Errorf("failed to persist intermediate batch of contract storage items and IDs: %w", seekErr)
+		go func() {
+			trieStore.Seek(storage.SeekRange{Prefix: []byte{byte(oldStoragePrefix)}}, func(k, v []byte) bool {
+				if seekErr.Load() != nil {
 					return false
 				}
-				batchCnt++
-				bc.log.Info("intermediate batch of contract storage items and IDs is persisted", zap.Int("batches persisted", batchCnt), zap.Duration("took", time.Since(p)), zap.Int("keys", keys))
-				p = time.Now()
-			}
-			// May safely omit KV copying.
-			k[0] = byte(newStoragePrefix)
-			cache.Store.Put(k, v)
-			cnt++
-			storageItmsCnt++
-
-			// @fixme: remove this part after #2702.
-			if bytes.HasPrefix(k[1:], mgmtContractPrefix) {
-				var hash util.Uint160
-				copy(hash[:], k[mgmtCSPrefixLen:])
-				err = stackitem.DeserializeConvertible(v, cs)
+				if cnt >= persistBatchSize {
+					workers.Inc()
+					cnt = 0
+					batchCnt++
+
+					batchStart := p
+					batchID := batchCnt
+					go func() {
+						k, err := cache.Persist()
```

Comment:
@roman-khimov, I can't find the bug here, but it exists. Concurrent persist logs sometimes show the following:
```
2022-11-21T14:42:35.081+0300	INFO	intermediate batch of contract storage items and IDs is persisted	{"batch ID": 3, "took": "31.96662271s", "keys": 199993}
2022-11-21T14:43:28.022+0300	INFO	persist took	{"time": "407.918899ms"}
2022-11-21T14:43:28.022+0300	INFO	intermediate batch of contract storage items and IDs is persisted	{"batch ID": 4, "took": "53.384117318s", "keys": 199999}
2022-11-21T14:44:23.882+0300	INFO	persist took	{"time": "355.836113ms"}
```
So it's less than 200K keys were persisted which is wrong. Here we used non-private DAO so changes inside persist should be handled with lock taken and I'd expect >= 200K keys to be persisted every time. Could you take a look, please?

Reply:
Single-threaded version works fine though.

Reply:
No woner, it's a concurrent process, it can wait for a while before going into `Persist` and a bit more KVs will fit into the current batch, but they will be counted for the next one, so you have a chance for the next batch to be smaller than 200K.

Reply:
I've refactored this code, now we have a single persisting routine.